### PR TITLE
Add Mac M1 publish runner

### PR DESCRIPTION
**Pull request by rstz**
IIUC, we only need to specify the correct runners to get the MacOS runner. Reference: https://github.blog/2023-10-02-introducing-the-new-apple-silicon-powered-m1-macos-larger-runner-for-github-actions/

@ianspektor Can you help me test this?

### CODE REVIEW

File: .github/workflows/publish.yaml [hunk 1]
```diff
@@ -54,11 +54,37 @@ jobs:
       - name: Save package for following jobs
         uses: actions/upload-artifact@v3
         with:
-          name: dist-macos-${{ matrix.python-version }}
+          name: dist-macos-intel-${{ matrix.python-version }}
+          path: dist
+
+  build-macos-arm:
+    runs-on: macos-latest-xlarge
```
Comment (ianspektor) [inline .github/workflows/publish.yaml:61]: is there a reason for using xlarge for arm and large for intel?

Comment (ianspektor) [inline .github/workflows/publish.yaml:61]: according to this https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories `latest` points to `macos-12`, but the one using M1 is `macos-14`?

Comment (rstz) [reply .github/workflows/publish.yaml:61]: As I understand Githubs documentation: large --> Intel. xlarge --> Arm https://docs.github.com/en/actions/using-github-hosted-runners/about-larger-runners/about-larger-runners#:~:text=The%20following%20machines%20sizes%20are%20available%20for%20macOS%20larger%20runners.

Comment (rstz) [reply .github/workflows/publish.yaml:61]: I think we can use the large machines, where IIUC correctly large and xlarge are how they are distinguished https://docs.github.com/en/actions/using-github-hosted-runners/about-larger-runners/about-larger-runners#:~:text=The%20following%20machines%20sizes%20are%20available%20for%20macOS%20larger%20runners.

Comment (ianspektor) [reply .github/workflows/publish.yaml:61]: Hm, agree, in the link I sent below I understood otherwise

**Comment by ianspektor** [comment]
@rstz ready on my side, review and merge away!

**Comment by rstz** [comment]
LGTM! I think you have to formally approve this

**Review by ianspektor — APPROVED** [review]

**Pull request merged**